### PR TITLE
Refactoring and adding convention, config and data.

### PR DESCRIPTION
**Pull request by MoMannn**

### CODE REVIEW

File: contracts/utils/Selector.sol [hunk 1]
```diff
@@ -3,7 +3,6 @@ pragma solidity ^0.4.23;
 import "../tokens/Xcert.sol";
 import "../tokens/BurnableXcert.sol";
 import "../tokens/PausableXcert.sol";
-import "../tokens/ChainableXcert.sol";
```
Comment (lknix) [inline contracts/utils/Selector.sol]: Commit message doesn't represent the changes in the commit

File: test/tokens/XcertMock.test.js [hunk 2]
```diff
@@ -26,13 +26,6 @@ contract('XcertMock', (accounts) => {
     await xcert.setPause(true);
     await assertRevert(xcert.transferFrom(accounts[0], accounts[1], id1));
   });
-
```
Comment (lknix) [inline test/tokens/XcertMock.test.js]: Same here - change the commit message please

File: test/tokens/XcertMock.test.js [hunk 3]
```diff
@@ -26,13 +26,6 @@ contract('XcertMock', (accounts) => {
...
-  it('correctly chains additional proof', async () => {
-    await xcert.chain(id1, mockProof2);
-    var proof = await xcert.tokenProof(id1);
-    assert.equal(proof, mockProof2);
-  });
-
```
Comment (lknix) [inline test/tokens/XcertMock.test.js:35]: https://github.com/0xcert/ethereum-xcert/pull/6/commits/37c9f2b0c0c03f4c486e43ad50c0a871806493a2, https://github.com/0xcert/ethereum-xcert/pull/6/commits/f5170c7c0c8958c858ba1ce31e6bf1612a89aea7 and this commit should be joined into one. What we want is that every change to the codebase (every commit) passes the tests. So if you selectively remove the code then some repo states become broken.

File: contracts/tokens/Xcert.sol [hunk 4]
```diff
@@ -13,6 +13,12 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
   using SafeMath for uint256;
   using AddressUtils for address;
 
+  /**
+   * @dev A bytes4 of keccak256 of json schema representing 0xcert protocol convention.
+   * @notice bytes4(keccak256(json)).
+   */
+  bytes4 private nftConvention;
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: I'd name this `nftConvetionId`.

File: contracts/tokens/Xcert.sol [hunk 5]
```diff
@@ -13,6 +13,12 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
   using SafeMath for uint256;
   using AddressUtils for address;
 
+  /**
+   * @dev A bytes4 of keccak256 of json schema representing 0xcert protocol convention.
+   * @notice bytes4(keccak256(json)).
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: `bytes4(keccak256(jsonSchema)).`

File: contracts/tokens/Xcert.sol [hunk 6]
```diff
@@ -13,6 +13,12 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
   using SafeMath for uint256;
   using AddressUtils for address;
 
+  /**
+   * @dev A bytes4 of keccak256 of json schema representing 0xcert protocol convention.
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: dev: Unique ID which determines each Xcert smart contract type by its JSON convention.
notice: calculated as `bytes4(keccak256(jsonSchema))`

File: contracts/tokens/Xcert.sol [hunk 7]
```diff
@@ -64,7 +64,6 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
     NFTokenMetadata(_name, _symbol)
     public
   {
-    require(_convention.length > 0);
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: Next time please split them into separate commits. One atomic change per commit.

File: contracts/utils/Selector.sol [hunk 8]
```diff
@@ -10,7 +10,7 @@ contract Selector {
     function calculateXcertSelector() public pure returns (bytes4) {
       Xcert i;
       return i.mint.selector
-         ^ i.tokenProof.selector
+      //   ^ i.tokenProof.selector
```
Comment (lknix) [inline contracts/utils/Selector.sol]: Let's not comment the code - if we don't need it, we remove it. Git history is our friend for all other cases.

File: contracts/tokens/Xcert.sol [hunk 9]
```diff
@@ -19,6 +19,21 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
    */
   bytes4 private nftConvention;
 
+  /*
+   * @dev Maps NFT ID to proof.
+   */
+  mapping (uint256 => string) internal idToProof;
+
+  /**
+   * @dev Map of protocol data.
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: Maybe: "Maps NFT ID to its on-chain protocol config."

File: contracts/tokens/Xcert.sol [hunk 10]
```diff
@@ -59,28 +74,37 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
...
   {
     nftConvention = _convention;
-    supportedInterfaces[0x355d09e9] = true; // Xcert
+    supportedInterfaces[0xe353dea7] = true; // Xcert
   }
 
-  /*
+  /**
    * @dev Mints a new NFT.
+   * @notice _dataKey and _dataValue length has to be the same.
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: what is this?

File: contracts/tokens/Xcert.sol [hunk 11]
```diff
@@ -59,28 +74,37 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
...
    * @param _to The address that will own the minted NFT.
    * @param _id The NFT to be minted by the msg.sender.
+   * @param _uri An URI pointing to NFT metadata.
    * @param _proof Cryptographic asset imprint.
-   * @param _uri An URI pointing to NFT metadata (optional, max length 2083).
+   * @param _config Array of config values.
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: "array of internal / protocol config values"? we should be as descriptive as possible here

File: contracts/tokens/Xcert.sol [hunk 12]
```diff
@@ -59,28 +74,37 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
...
+   * @param _data Array of convention data values.
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: Array of on-chain convention data values.

File: contracts/tokens/Xcert.sol [hunk 13]
```diff
@@ -94,6 +118,68 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
     _convention = nftConvention;
   }
 
+  /**
+   * @dev Gets proof for NFT token.
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: Returns the proof of NFT token.

File: contracts/tokens/Xcert.sol [hunk 14]
```diff
@@ -94,6 +118,68 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
...
+    validNFToken(_tokenId)
+    external
+    view
+    returns(string)
+  {
+    return idToProof[_tokenId];
+  }
+
+  /**
+   * @dev Sets value for specified key.
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: Modify on-chain convention data by setting a new value for a given index field.

File: contracts/tokens/Xcert.sol [hunk 15]
```diff
@@ -94,6 +118,68 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
...
+    onlyOwner
+    validNFToken(_tokenId)
+    external
+  {
+    require(_index < data[_tokenId].length);
+    data[_tokenId][_index] = _value;
+  }
+
+  /**
+   * @dev Gets value of key for specific NFT.
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: Returns on-chain convention data value for a given index field.

File: contracts/tokens/Xcert.sol [hunk 16]
```diff
@@ -94,6 +118,68 @@ contract Xcert is NFTokenEnumerable, NFTokenMetadata {
...
+    public
+    view
+    returns(bytes32 value)
+  {
+    require(_index < data[_tokenId].length);
+    value = data[_tokenId][_index];
+  }
+
+  /*
+   * @dev Gets expiration date from config values.
```
Comment (lknix) [inline contracts/tokens/Xcert.sol]: Returns expiration date from on-chain NFT config.

File: contracts/utils/Selector.sol [hunk 17]
```diff
@@ -10,7 +10,11 @@ contract Selector {
     function calculateXcertSelector() public pure returns (bytes4) {
       Xcert i;
       return i.mint.selector
-      //   ^ i.tokenProof.selector
```
Comment (lknix) [inline contracts/utils/Selector.sol]: This crept in.

File: test/tokens/BurnableXcert.test.js [hunk 18]
```diff
@@ -8,29 +9,31 @@ contract('BurnableXcert', (accounts) => {
   let id2 = web3.sha3('test2');
   let id3 = web3.sha3('test3');
   let id4 = web3.sha3('test4');
-  let mockProof = "1e205550c271490347e5e2393a02e94d284bbe9903f023ba098355b8d75974c8";
+  const proof = '1e205550c271490347e5e2393a02e94d284bbe9903f023ba098355b8d75974c8';
+  let config = [web3Util.padLeft(web3Util.numberToHex(1821195657), 64)];
```
Comment (lknix) [inline test/tokens/BurnableXcert.test.js]: How did you get the 182119565?

File: test/tokens/Xcert.test.js [hunk 19]
```diff
@@ -286,12 +289,47 @@ contract('Xcert', (accounts) => {
...
+    await xcert.mint(accounts[1], id2, 'url2', proof, config, data);
     const tokenURI = await xcert.tokenURI(id2);
     assert.equal(tokenURI, 'url2');
   });
 
   it('throws when trying to get uri of none existant NFToken id', async () => {
     await assertRevert(xcert.tokenURI(id4));
   });
+
+  it('returns the correct proof for NFToken id', async () => {
```
Comment (lknix) [inline test/tokens/Xcert.test.js]: Add a test for missing NFT id

File: test/tokens/Xcert.test.js [hunk 20]
```diff
@@ -286,12 +289,47 @@ contract('Xcert', (accounts) => {
...
+    await xcert.mint(accounts[1], id2, 'url2', proof, config, data);
+    let tokenProof = await xcert.tokenProof(id2);
+    assert.equal(proof, tokenProof);
+  });
+
+  it('correctly checks expiration date for NFT', async () => {
+    await xcert.mint(accounts[1], id2, 'url2', proof, config, data);
+    let expirationDate = await xcert.tokenExpirationDate.call(id2);
+    assert.equal(expirationDate, config[0]);
```
Comment (lknix) [inline test/tokens/Xcert.test.js]: same here

File: test/tokens/Xcert.test.js (hunk 19 above)
Comment (MoMannn) [reply test/tokens/Xcert.test.js]: Why? It has a modifier validNFToken which has been already tested.

File: test/tokens/BurnableXcert.test.js (hunk 18 above)
Comment (lknix) [reply test/tokens/BurnableXcert.test.js]: Ah, ok it's a timestamp

File: test/tokens/Xcert.test.js (hunk 19 above)
Comment (lknix) [reply test/tokens/Xcert.test.js]: Ah, ok, I might have been too quick.

File: contracts/tokens/Xcert.sol (hunk 9 above)
Comment (lknix) [reply contracts/tokens/Xcert.sol]: You're right. No need to use the term, just sounded good when I wrote it :)

File: contracts/tokens/Xcert.sol (hunk 4 above)
Comment (xpepermint) [reply contracts/tokens/Xcert.sol]: I agree.

File: test/tokens/BurnableXcert.test.js (hunk 18 above)
Comment (xpepermint) [reply test/tokens/BurnableXcert.test.js]: Use `cost` instead of `let` when possible.

**Comment by lknix** [comment]
@MoMannn if you've addressed all my comments (either with commit changes or replies about why not to do it), then LGTM!
